java/algorithm/src/main/java/org/apache/arrow/algorithm/search/ParallelSearcher.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.search;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+
+import org.apache.arrow.vector.ValueVector;
+import org.apache.arrow.vector.compare.Range;
+import org.apache.arrow.vector.compare.RangeEqualsVisitor;
+
+/**
+ * Search for a value in the vector by multiple threads.
+ * This is often used in scenarios where the vector is large or
+ * low response time is required.
```

Comment:
It could be helpful if this linked to the single-threaded `VectorSearcher` as an alternative (and if that linked here as well). It should also be clear about the contract of this class - while `VectorSearcher#linearSearch` will find the _first_ matching value, this will only find _a_ matching value.

Reply:
I agree with you it is a good idea to link to VectorSearcher#linearSearch to make the contract clear. 

We give an independent implementation here, mainly because it is impossible to quit the search early, if another thread has found a match.

Reply:
Could you just add a line to this docstring outlining the difference? Something to the effect of "If there are multiple instances of the key in the target vector, this makes no guarantees about which instance is returned. For an alternative search implementation that always finds the first instance of the key, see `VectorSearcher#linearSearch`."

Reply:
I have updated the JavaDoc according to your suggestion. Thanks a lot for the good point.